File: 0-数据结构/4-树与二叉树/二叉搜索树(BST)/二叉搜索树定义.cpp
#include <iostream>
using namespace std;

// 定义节点结构
struct Node {
    int data;
    Node *left, *right;
    Node(int val) : data(val), left(nullptr), right(nullptr) {}
};

class BST {
public:
    Node* root = nullptr;

    // 插入节点
    void insert(int val) {
        root = insertRecursive(root, val);
    }

    // 中序遍历 (结果应为有序序列)
    void inorder() {
        inorderRecursive(root);
        cout << endl;
    }

private:
    Node* insertRecursive(Node* node, int val) {
        if (node == nullptr) return new Node(val);

        if (val < node->data)
            node->left = insertRecursive(node->left, val);
        else if (val > node->data)
            node->right = insertRecursive(node->right, val);
            
        return node;
    }

    void inorderRecursive(Node* node) {
        if (node) {
            inorderRecursive(node->left);
            cout << node->data << " ";
            inorderRecursive(node->right);
        }
    }
};

int main() {
    BST tree;
    int values[] = {50, 30, 20, 40, 70, 60, 80};
    for (int x : values) tree.insert(x);

    cout << "C++ BST 中序遍历结果: ";
    tree.inorder();
    return 0;
}

File: 0-数据结构/4-树与二叉树/二叉搜索树(BST)/二叉搜索树定义.py
class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

class BST:
    def __init__(self):
        self.root = None

    def insert(self, key):
        if self.root is None:
            self.root = Node(key)
        else:
            self._insert_recursive(self.root, key)

    def _insert_recursive(self, node, key):
        if key < node.key:
            if node.left is None:
                node.left = Node(key)
            else:
                self._insert_recursive(node.left, key)
        elif key > node.key:
            if node.right is None:
                node.right = Node(key)
            else:
                self._insert_recursive(node.right, key)

    def inorder_traversal(self, node, result):
        if node:
            self.inorder_traversal(node.left, result)
            result.append(node.key)
            self.inorder_traversal(node.right, result)
        return result

# 使用示例
bst = BST()
for val in [50, 30, 20, 40, 70, 60, 80]:
    bst.insert(val)

print("Python BST 中序遍历结果:", bst.inorder_traversal(bst.root, []))